gpytorch/kernels/kernel.py

```diff
@@ -304,11 +331,7 @@ def _covar_dist(
         if diag:
             # Special case the diagonal because we can return all zeros most of the time.
             if x1_eq_x2:
-                # We are computing distances between points and themselves, which are all zero
-                if batch_dims == (0, 2):
-                    res = torch.zeros(x1.shape[0] * x1.shape[-1], x2.shape[-2], dtype=x1.dtype, device=x1.device)
-                else:
-                    res = torch.zeros(x1.shape[0], x2.shape[-2], dtype=x1.dtype, device=x1.device)
+                res = torch.zeros(*x1.shape[:-2], x1.shape[-2], dtype=x1.dtype, device=x1.device)
```

Comment:
Does this break the `batch_dims` case?

Reply:
No, by this point the d dimension is already its own batch dimension (one of the changes of this PR is to handle batch_dims with multibatch instead of combining it with another batch dimension.